File: src/main/java/com/assets/board/service/ExchangeRateService.java
package com.assets.board.service;

import com.assets.board.client.NbuDataClient;
import com.assets.board.entity.ExchangeRate;
import com.assets.board.repository.ExchangeRateRepository;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;

@Slf4j
@Service
@AllArgsConstructor
public class ExchangeRateService {

    private final ExchangeRateRepository repository;
    private final NbuDataClient nbuDataClient;

    @Transactional
    public BigDecimal getRateForDate(LocalDate date) {
        return repository.findByDate(date)
                .map(ExchangeRate::getNbuRate)
                .orElseGet(() -> fetchAndSaveRate(date));
    }

    @Transactional
    public ExchangeRate saveExchangeRate(LocalDate date, BigDecimal rate) {
        ExchangeRate exchangeRate = new ExchangeRate();
        exchangeRate.setDate(date);
        exchangeRate.setNbuRate(rate);

        return repository.save(exchangeRate);
    }

    private BigDecimal fetchAndSaveRate(LocalDate date) {
        BigDecimal exchangeRate = nbuDataClient.exchangeRate(date);
        ExchangeRate rate = this.saveExchangeRate(date, exchangeRate);
        return rate.getNbuRate();
    }
}

File: src/main/java/com/assets/board/service/impl/ExchangeRateService.kt
package com.assets.board.service.impl

import com.assets.board.client.NbuDataClient
import com.assets.board.entity.ExchangeRate
import com.assets.board.repository.ExchangeRateRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDate

@Service
@Transactional
class ExchangeRateService(
    private val repository: ExchangeRateRepository,
    private val nbuDataClient: NbuDataClient
) {
    fun getRateForDate(date: LocalDate?): BigDecimal? {
        return repository.findByDate(date)?.nbuRate
            ?: date?.let { fetchAndSaveRate(it) }
    }

    fun saveExchangeRate(date: LocalDate, rate: BigDecimal): ExchangeRate {
        val exchangeRate = ExchangeRate(
            date = date,
            nbuRate = rate
        )
        return repository.save(exchangeRate)
    }

    private fun fetchAndSaveRate(date: LocalDate): BigDecimal? {
        val exchangeRate = nbuDataClient.exchangeRate(date)
        return saveExchangeRate(date, exchangeRate).nbuRate
    }
}
